stage1: remove --pid-file patch from systemd-nspawn

alban opened the pull request:
Instead of asking systemd-nspawn to write its "pid" file, rkt can write
the parent pid "ppid". Then, update getPID() to get the pid file from
/proc/$ppid/task/$ppid/children. In this way, we don't need to patch
systemd-nspawn.

getPID() is able to read both the "pid" file and the "ppid" file, so the
command "rkt status" or "rkt enter" will work when attempted against a
pod started with an older version of stage1.

See also #885.

iaguis commented:
It doesn't work for me.

It seems `/proc/$pid/task/$pid/children` needs `CONFIG_CHECKPOINT_RESTORE` enabled in the kernel. Unfortunately it's [disabled in my distro (Arch)](https://bugs.archlinux.org/task/43088) because it depends on `CONFIG_EXPERT` :disappointed:

I commented on the (relevant archlinux bug)[https://bugs.archlinux.org/task/43088#comment135546] to try to get it enabled.

jonboulle commented on rkt/pods.go:
```diff
@@ -733,14 +734,33 @@ func (p *pod) getState() string {
 func (p *pod) getPID() (pid int, err error) {
+	// check the "pid" (process ID) file if it exists, and on "ppid"
+	// (parent process ID). Recent stage1 uses "ppid" but we still
+	// support older rkt versions or alternative stage1 implementations.
```
We gotta define this ABI more explicitly as part of #572

alban replied:
Ok. I should add some documentation in the [Stage1 ACI implementor's guide](https://github.com/coreos/rkt/blob/master/Documentation/devel/stage1-implementors-guide.md#rkt-enter--coreoscomrktstage1enter). At the moment, it is out of sync since #826.

alban replied:
I am thinking of reverting a part the patch in #826 to read the pid files in the `enter` entrypoint rather than in stage0. My idea is that the pid is an implementation detail of stage1 and should not leak in stage0: other stage1 such as ones based on KVM might not have a process leader for the container. In this way, the pid files would not need to be specified in the [Stage1 ACI implementor's guide](https://github.com/coreos/rkt/blob/master/Documentation/devel/stage1-implementors-guide.md#rkt-enter--coreoscomrktstage1enter).

alban commented:
Is it acceptable to break support for kernels that don't have `CONFIG_CHECKPOINT_RESTORE`? It is enabled in Debian, Fedora, Ubuntu, CoreOS, etc. but not Arch.

As Iago said, this option is strongly recommended for running systemd and rkt definitely needs some kernel options (see #918) but I am not sure what to do for Arch users...

alban commented:
I rebased this branch after #919 got merged. I should still address the issues mentioned above (support for Arch? ABI?)

jonboulle commented:
My inclination is that it might make sense to block this on being in Arch (as a somewhat arbitrary benchmark for its ubiquity). Is it worth pinging kernel to see if it's still strictly necessary to gate this behind `CONFIG_CHECKPOINT_RESTORE` ?

alban commented:
> Is it worth pinging kernel to see if it's still strictly necessary to gate this behind `CONFIG_CHECKPOINT_RESTORE` ?

I asked them:
http://permalink.gmane.org/gmane.linux.kernel.api/11370

alban commented:
@vcaputo @yifan-gu : I rebased this branch and added a new commit which changes the logic you implemented in #826 (race on pod.getPID). I am trying to move the pid file code back into stage1 so that it would be an implementation detail of stage1 and it would not need to be described in the [Stage1 ACI implementor's guide](https://github.com/coreos/rkt/blob/master/Documentation/devel/stage1-implementors-guide.md#rkt-enter--coreoscomrktstage1enter). Other stage1 such as ones based on KVM might not have a process leader for the container.

There are a couple of TODO in my commit that I have not yet addressed, but I think the idea is here.

Did you have any scenario to trigger the bug addressed by #826? I described 3 scenarios in the commit log (https://github.com/endocode/rkt/commit/a39337b41302ef279feeef1932f348311bb8980a) but I don't know if they are related to the one you had.

alban commented:
@iaguis wrote a kernel patch to introduce `CONFIG_PROC_CHILDREN` which enables `/proc/pid/task/tid/children`, so `CONFIG_EXPERT` and `CONFIG_CHECKPOINT_RESTORE` will not be necessary. The [patch has been added to the -mm tree](http://ozlabs.org/~akpm/mmots/broken-out/fs-proc-introduce-config_proc_children.patch) and will go to linux-next.

yifan-gu commented:
@alban Sorry, missed the PR for several days. The original scenario is that when `rkt enter` report `no pid file` when it enters soon after `rkt run`. I think this is covered by your Test1.

vcaputo commented:
@alban I couldn't (and can't) imagine any stage1 implementation which will not have a unix process to which we can attach the container's lock fd to.  It really seems to belong in stage0 IMHO, even kvm instances will have a user process.

iaguis commented on rkt/pods.go:
```diff
@@ -731,16 +733,87 @@ func (p *pod) getState() string {
+	_, err := os.Stat("/proc/1/task/1/children")
+	if err == nil {
+		var b []byte
+		b, err = ioutil.ReadFile(fmt.Sprintf("/proc/%d/task/%d/children", ppid, ppid))
```
you can just assign with `:=` here

alban replied:
fixed.

iaguis commented on rkt/pods.go:
```diff
@@ -731,16 +733,87 @@ func (p *pod) getState() string {
+			if len(children) == 2 && children[1] != "" {
+				return -1, fmt.Errorf("Too many children of pid %d", ppid)
+			}
+			_, err = fmt.Sscanf(children[0], "%d ", &pid)
```
same here

iaguis commented on rkt/pods.go:
```diff
@@ -731,16 +733,87 @@ func (p *pod) getState() string {
+		if len(statFields) != 5 {
+			return -1, fmt.Errorf("Incomplete file %q", filename)
+		}
+		if statFields[3] == fmt.Sprintf("%d", ppid) {
```
you used FormatInt before and now you use Sprintf, choose one

alban replied:
Fixed with Sprintf.

iaguis commented on scripts/debug/stage1_install_busybox.sh:
```diff
@@ -52,5 +52,6 @@ for link in ${BUSYBOX_LINKS}; do
 echo "Busybox installed. Use the following command to enter pod's stage1:"
-SYSTEMD_PID=$(sudo cat "${RKT_RUN_DIR}/${UUID}/pid")
+SYSTEMD_PPID=$(sudo cat "${RKT_RUN_DIR}/${UUID}/ppid")
+SYSTEMD_PID=$(sudo cat /proc/$SYSTEMD_PID/task/$SYSTEMD_PID/children)
```
Hmmm... this doesn't make sense

iaguis commented on Documentation/devel/stage1-implementors-guide.md:
```diff
@@ -38,6 +38,10 @@ In the bundled rkt stage 1 which includes systemd-nspawn and systemd, the entryp
+The resolved entrypoint must inform rkt of its pid for the benefit of `rkt enter`. rkt supports two ways of doing that, stage1 implementors must pick one:
+
+1. Writing the pid of the container's PID 1 in the file "/var/lib/rkt/pods/$uuid/pid"
+2. Writing the parent pid of the container's PID 1 in the file "/var/lib/rkt/pods/$uuid/ppid". This assumes it has only one child.
```
This is a bit confusing. What about

1. Writing to "/var/lib/rkt/pods/$uuid/pid" the PID of the process that is PID 1 in the container
2. Writing to "/var/lib/rkt/pods/$uuid/ppid" the PID of the parent of the process that is PID 1 in the container. This assumes it has only one child.

iaguis commented on stage1/init/init.go:
```diff
@@ -381,6 +375,20 @@ func stage1() int {
+		fmt.Fprintf(os.Stderr, "Cannot get current working directory: %v\n", err)
+		return 4
+	}
+	err = ioutil.WriteFile(filepath.Join(out, "ppid"),
```
Even though you describe it in "Documentation/devel/stage1-implementors-guide.md", it's confusing seeing "ppid" and then `os.Getpid()`, maybe add a comment like

"we are the parent of the process that is PID 1 in the container so we write our PID to "ppid""

iaguis commented on Documentation/devel/stage1-implementors-guide.md:
```diff
@@ -57,9 +61,10 @@ An alternative stage 1 would need to do whatever is appropriate for entering the
-1. image id of the specific application to enter
-2. cmd to execute
-3. any cmd arguments
+1. pid of the container's PID 1. rkt finds that pid by one of the two supported methods described in the `rkt run` section.
```
"PID of the process that is PID 1 in the container"

Also, in general you use "PID" (uppercase) when you refer to "PID 1" and "pid" otherwise, I think you should always use PID and remove that distinction

iaguis commented on rkt/pods.go:
```diff
@@ -731,16 +733,85 @@ func (p *pod) getState() string {
+		if err == nil {
+			children := strings.SplitN(string(b), " ", 2)
+			if len(children) == 2 && children[1] != "" {
+				return -1, fmt.Errorf("Too many children of pid %d", ppid)
```
Too -> too

iaguis commented on rkt/pods.go:
```diff
@@ -731,16 +733,85 @@ func (p *pod) getState() string {
+		}
+		statFields := strings.SplitN(string(statBytes), " ", 5)
+		if len(statFields) != 5 {
+			return -1, fmt.Errorf("Incomplete file %q", filename)
```
Incomplete -> incomplete

alban commented:
@yifan-gu: the new tests added in #972 and #973 should cover it now.

@vcaputo: ok, I changed it back and updated the documentation. PTAL.

@iaguis: I rebased and updated the branch after your review.

iaguis commented:
LGTM.

I tested it both on a kernel with and without support for `/proc/<pid>/tasks/<tid>/children` and it worked.

I would like another review before merging it though.

vcaputo commented on rkt/pods.go:
```diff
@@ -731,16 +733,85 @@ func (p *pod) getState() string {
 }
 
+// Returns the pid of the child, or 0 if not ready
+func getChildPID(ppid int) (int, error) {
```
nit:
Overloading the pid return value to indicate when the child can't be read (pid of 0) would be normal in C, but in Go you should probably be returning a specific error.  Then the caller will be self-documenting in specifically handling that error.

vcaputo commented on rkt/pods.go:
```diff
@@ -731,16 +733,85 @@ func (p *pod) getState() string {
+		if err != nil {
+			return -1, err
+		}
+		if len(fi) == 0 {
```
This test is unnecessary:

>     If n > 0, Readdir returns at most n FileInfo structures. In this case,
    if Readdir returns an empty slice, it will return a non-nil error
    explaining why. At the end of a directory, the error is io.EOF.

vcaputo commented:
small things, otherwise lgtm

alban commented:
Thanks for the review.

I added a fixup commit, for easier review. If it's fine, I will squash that fixup commit before merging.

alban commented:
Semaphore is green and the tests also pass on my laptop.

@iaguis reviewed it out of band.
